### PR TITLE
Add fsfreeze functionality to snapshot

### PR DESCRIPTION
**Pull request by ejweber**
#### Which issue(s) this PR fixes:

longhorn/longhorn#2187

#### What this PR does / why we need it:

- Attempt to bind mount the file system on the root partition of a Longhorn volume and freeze it with fsfreeze if requested by snapshot command line argument or snapshot gRPC command. Otherwise, just sync like we have always done.
- Attempt to unfreeze the file system on the root partition during frontend shutdown in case we are in the middle of a snapshot.
- Attempt to unfreeze the file system on the root partion of a Longhorn volume during frontend startup in case its engine previously crashed before an unfreeze was issued.

#### Additional documentation or context

Depends on https://github.com/longhorn/types/pull/8.

We need to roll back the debugging commit and fix the imports before merging.

### CODE REVIEW

**Comment by ejweber** [comment]
I found a pretty convenient way to modify the integration test that was written for this feature's original implementation to work with the current implementation. It tests basic functionality (e.g., when `Snapshot` is called with `shouldFreeze == true`, we can successfully detect the mounted file system, do the bind mound, and attempt a freeze).

Otherwise, I manually tested the following scenarios using this modified engine and the modified instance manager. In the interest of time, I write up an exhaustive description of how the tests were executed, but here are the basic commands involved:

- Run a simple Longhorn volume in a Docker container.
  `docker run --privileged -v /:/host --rm --net host --name test longhornio/longhorn-engine:2149c73d launch-simple-longhorn test 10g tgt-blockdev`
- Make a file system on the volume and mount it.
  `mkfs.ext4 /dev/longhorn/test`
  `mount /dev/longhorn/test /mnt/test`
- Constantly write data to the file system while snapshotting, etc.
  `while true; do dd of=/mnt/test/file if=/dev/urandom bs=16M; done`
- Watch the dirty page cache while freeze is ongoing.
  `while true; do cat /proc/meminfo | grep -i dirty; sleep 0.5; done`
- Check inside the Docker container for mounted volumes.
  `docker exec test bash -c " while true; do mount | grep test; sleep 0.5; done"`
- Create a snapshot with or without a request to freeze.
  `bin/longhorn -url localhost:10015 snapshot create --freeze-fs`
  `bin/longhorn -url localhost:10015 snapshot create`
- Check for interesting processes and determine if they are stuck and why.
  `while true; do ps -eo pid,ppid,pgid,user,stat,pcpu,comm,wchan:32 | grep -e freeze -e dmsetup -e " dd" -e laun; sleep 1; done`
- Delete and recreate Longhorn processes using instance-manager.
  `docker exec test longhorn-instance-manager process delete --name test-e`
  `docker exec test longhorn-instance-manager process delete --name test-r`
  `docker exec test longhorn-instance-manager process create --name test-r --binary /usr/local/bin/longhorn --port-count 15 --port-args --listen,localhost: -- replica /volume/ --size 10g`
  `docker exec test longhorn-instance-manager process create --name test-e --binary /usr/local/bin/longhorn --port-count 1 --port-args --listen,localhost: -- controller test --frontend tgt-blockdev --size 10g --current-size 10g --replica tcp://localhost:10000`
- Kill the whole Longhorn volume immediately.
  `kill -9 -<process_group of the launch process>`

Here are the test cases. All tests were executed on a known "good" kernel (Rocky 9.3 with an upstream `6.8.7-1.el9.elrepo.x86_64` kernel installed).

---

Take a snapshot without --freeze-fs.

Check:
  Logs clearly indicate system sync.
  Dirty pages trend toward zero before the snapshot is finished (but don't necessarily reach it).

---

Take a snapshot with --freeze-fs.
  
Check:
  Logs clearly indicate freeze and unfreeze.
  Logs do not indicate system sync.
  Loop running mount command clearly shows mount being created and destroyed.
  Loop running cat command clearly show cache being flushed.

---

Take a snapshot with --freeze-fs. Ctrl+C the instance manager container while the freeze is ongoing.

Check:
  Logs clearly indicate freeze.
  Logs indicate a timeout unfreezing. (This is because we automatically try to unfreeze during shutdown, but unfreeze is blocked by freeze. We return anyway.)
  Logs may indicate a later attempt to unfreeze. (This is because our freeze attempt finally fails when I/O errors return and then we attempt to undo it. It's really not particularly helpful or interesting, but it is expected.)
  The snapshot fails with an indication that freeze failed.
  The dd command gets unstuck (once I/O errors hit after the iSCSI timeout).
  The file system can be unmounted.

---

Take a snapshot with --freeze-fs. kill -9 the instance manager container while the freeze is ongoing.

Check:
  Logs clearly indicate freeze.
  The snapshot fails with an error that includes an EOF.
  The dd command gets unstuck (once I/O errors hit after the iSCSI timeout).
  The file system can be unmounted.

---

Take a snapshot with --freeze-fs. kill -9 the instance manager container BETWEEN FREEZE AND LOCK (need code modification).

Check:
  Logs clearly indicate freeze.
  The snapshot fails with an error that includes an EOF.
  After instance-manager is restarted:
    Logs clearly indicate unfreeze (from the instance-manager startup logic).
	The dd command gets unstuck.
	The file system can be unmounted.

---

Take a snapshot with --freeze-fs. Stop the engine using a call to instance-manager BETWEEN FREEZE AND LOCK (need code modification).

Check:
  Logs clearly indicate freeze.
  The snapshot fails with an error that includes an EOF.
  Logs clearly indicate unfreeze (from the shutdown frontend logic).
  The dd command gets unstuck.
  The file system can be unmounted.

---

Take a snapshot with --freeze-fs. kill -9 the engine process (but not instance-manager) BETWEEN FREEZE AND LOCK (need code modification).
  Logs clearly indicate freeze.
  The snapshot fails with an error that includes an EOF.
  To recover:
    Use a call to instance-manager to delete the engine and replica processes.
	Use a call to instance-manager to start the engine and replica processes.
  After this:
    Logs clearly indicate unfreeze (from the startup frontend logic).
    The dd command gets unstuck.
    The file system can be unmounted.  

  One concern here. Should we do this earlier? If not, we can never issue an unfreeze for an engine that can't get at least one backend running somewhere.

---

Take a snapshot with --freeze-fs. Then, quickly take another snapshot.
    Logs clearly indicate freeze and unfreeze for first snapshot.
	The snapshot fails with an indication that the file system is already being frozen

File: pkg/controller/control.go
Comment (PhanLe1010) [inline pkg/controller/control.go:1382]: Should we directly find the filesystem on the root partition of the `devicePath` and mount it to `freezePoint` so we don't have to find and bind mount the existing mount points? And if there is no filesystem on the root partition, we can do no-op and fallback to sync.

By doing this, we don't have to worry that the upper layer is unmounting while we are doing bind mount and attempt to freeze

File: pkg/frontend/tgt/frontend.go
Comment (PhanLe1010) [inline pkg/frontend/tgt/frontend.go:173]: Should we prioritize the mount point we created first (the one at `freezePointDirectory  = "/var/lib/longhorn/freeze"`)?

File: pkg/frontend/tgt/frontend.go
Comment (PhanLe1010) [inline pkg/frontend/tgt/frontend.go:87]: Should we put this logic into [func (d *LonghornDevice) Shutdown()](https://github.com/longhorn/longhorn-engine/blob/a807f0fd6bfd9c4700f2c19808038e87a2ab814e/vendor/github.com/longhorn/go-iscsi-helper/longhorndev/dev.go#L177) instead?

File: pkg/frontend/tgt/frontend.go
Comment (PhanLe1010) [inline pkg/frontend/tgt/frontend.go:79]: Should we put this logic into [func (d *LonghornDevice) startScsiDevice(](https://github.com/longhorn/longhorn-engine/blob/a807f0fd6bfd9c4700f2c19808038e87a2ab814e/vendor/github.com/longhorn/go-iscsi-helper/longhorndev/dev.go#L122)

File: pkg/controller/control.go
Comment (ejweber) [reply pkg/controller/control.go:1382]: I thought about this, and I don't feel too strongly either way. If a different implementation did as you suggested, it would probably be fine to me. Here are a couple reasons the bind mount approach might be better.

- If the file system is not mounted, freezing it is unnecessary. The bind mount approach only applies to a mounted file system.
- I'm not 100% sure of the behavior when directly mounting a file system that is already mounted. I tried it for ext4 and XFS volumes in my Rocky environment and it seemed to have the same effect as a bind mount. However, CSI drivers (including ours) always bind mount the staging path to the target path instead of doing a direct mount to the target path. It seems safer to me to follow suit.

**Comment by ejweber** [comment]
Reworking some things in response to @PhanLe1010's feedback. Will reopen again when I am back to fairly confident.

File: pkg/frontend/tgt/frontend.go
Comment (ejweber) [reply pkg/frontend/tgt/frontend.go:173]: It probably wouldn't hurt. I'll add this functionality in an upcoming pass.

File: pkg/frontend/tgt/frontend.go
Comment (ejweber) [reply pkg/frontend/tgt/frontend.go]: While moving the logic into `go-iscsi-helper`, I realized that we should almost certainly execute it BEFORE `startFrontend`. If there are no available backends, we will never reach `startFrontend`. That's probably fine for the existing cleanup `startFrontend` might do (it'll strand a block device on the node), but it's NOT fine for the unfreeze case. We need to make sure the filesystem gets unfrozen, regardless of whether or not the engine can run.

As a result, I'm keeping the logic in `longhorn-engine`, but moving it to the `util` package so it can be executed earlier.

File: pkg/frontend/tgt/frontend.go
Comment (ejweber) [reply pkg/frontend/tgt/frontend.go]: See my reply to your similar comment about `startScsiDevice`. I will keep the logic in `longhorn-engine` because of the new location we will execute it from.

File: pkg/controller/control.go
Comment (ejweber) [reply pkg/controller/control.go:249]: We discussed this offline. While there are some pros and cons to both behaviors, for now we will leave it as a `TryLock`. It is mainly because `Lock` may be stuck forever in some unknown scenario, but the `Snapshot` client will give up waiting. In that case, the user will not know why the snapshot failed and the goroutine will exist indefinitely. It may be better to just tell the user the snapshot failed immediately and allow them to retry. 

FWIW, the argument for `Lock` is quite valid as well. We can almost certainly get the lock if we are patient, so we don't really need to return failure. If we notice that the chosen behavior is causing problems, we can reevaluate.

**Comment by ejweber** [comment]
Moved back to ready for review @PhanLe1010. Sorry for the delay. I responded to all of your comments and made changes with respect to some of them.

I also fixed an issue Codefactor had with my integration test and ensured we actually cleaned up the directories within the instance-manager container we bind mounted to (in case a long running instance-manager used too many!).

**Comment by PhanLe1010** [comment]
> ensured we actually cleaned up the directories within the instance-manager container we bind mounted to (in case a long running instance-manager used too many!)

Could you point me to this? Thank you

**Comment by ejweber** [comment]
Sorry for the confusion. [This commit](https://github.com/longhorn/longhorn-engine/pull/1083/commits/d57877e1f0252cd07fd10ecc276b26ce3944b0cc) uses CleanupMountPoint in two places:

- One where we previously just used Unmount (so did not clean up the directory).
- One where we previously did not Unmount at all. (This was an oversight that could have led to mounts for shut down engines accumulating over time.)

**Review by PhanLe1010 — DISMISSED** [review]
LGTM

Thank you for the great implementation!

File: Dockerfile.dapper
Comment (james-munson) [inline Dockerfile.dapper:16]: Should be "-v /tmp:/host/tmp" in the comment (note missing "/").

File: pkg/controller/control.go
Comment (james-munson) [inline pkg/controller/control.go:1387]: "save" -> "safe"

**Review by james-munson — APPROVED** [review]
Just a couple of nits in comments, no issues.  LGTM

**Comment by ejweber** [comment]
Since https://github.com/longhorn/types/pull/8 merged and I have two approving reviews, I will remove my debug commits and rebase in preparation for merge soon.

File: Dockerfile.dapper
Comment (shuo-wu) [inline Dockerfile.dapper:121]: Remember to remove this commit when the PR is ready to go

File: pkg/frontend/tgt/frontend.go
Comment (shuo-wu) [inline pkg/frontend/tgt/frontend.go:81]: `"..., will continue the tgt shutdown"`

File: pkg/util/fsfreeze.go
Comment (shuo-wu) [inline pkg/util/fsfreeze.go]: Should these functions return a non-empty value (`/var/lib/longhorn/freeze/`) when the input is empty?

File: pkg/controller/control.go
Comment (ejweber) [reply pkg/controller/control.go]: I'm assuming we agree that this does not matter (since unmounting the source mount will not unmount the bind mount). I'm adding a comment to this effect.

File: pkg/controller/control.go
Comment (ejweber) [reply pkg/controller/control.go]: It looks like this comment is on outdated code from a previous commit. We already moved to using `mount.CleanUpMountPoint` instead of `mounter.Unmount` at all locations.

File: pkg/util/fsfreeze.go
Comment (ejweber) [reply pkg/util/fsfreeze.go]: If I understand you correctly, you are saying it is safer if we return an empty value instead of whatever the base of the join is? Makes sense to me.

**Comment by ejweber** [comment]
Added https://github.com/longhorn/longhorn-engine/pull/1083/commits/66fa2e6855196c400b831e36c0da8174b9e86181 for @shuo-wu's comments, removed debugging helpers, and pointed back to upstream types. Rebase is a mess given the number of commits / time in development and the dependency changes since I started, so we can use squash instead. Ready to go pending @shuo-wu's approval.

**Review by shuo-wu — APPROVED** [review]

**Comment by shuo-wu** [comment]
After resolving the rebase and conflicts, it's good to go

**Pull request merged**
